extras/create_snapshots_reports_scorecard.py

```diff
@@ -215,49 +225,147 @@ def create_weekly_snapshots(db, cyhy_db_section):
+            # method of finding all descendants
+            org_descendants = db.RequestDoc.get_all_descendants(org_id)
+
+    if snapshot_process.returncode == 0:
+        logging.info("Successful snapshot: %s (%.2f s))", org_id, snapshot_duration)
+        with ss_lock:
+            successful_snapshots.append(org_id)
+            if org_descendants and not use_only_existing_snapshots:
+                logging.info(
+                    " - Includes successful descendant snapshot(s): %s", org_descendants
+                )
+                successful_snapshots.extend(org_descendants)
+    else:
+        logging.error("Unsuccessful snapshot: %s", org_id)
+        with fs_lock:
+            failed_snapshots.append(org_id)
+            if org_descendants and not use_only_existing_snapshots:
+                logging.error(
+                    " - Unsuccessful descendant snapshot(s): %s", org_descendants,
+                )
+                failed_snapshots.extend(org_descendants)
+        logging.error("Stderr failure detail: %s %s", data, err)
+    return snapshot_process.returncode
+
+
+def create_snapshots(org_list, db, cyhy_db_section):
+    """Create a snapshot for each organization in a list."""
+    for org_id in org_list:
+        logging.info(
+            "[%s] Starting snapshot for: %s", threading.current_thread().name, org_id
         )
-        data, err = output.communicate("yes")
-        snap_time = time.time() - snap_time
-        longest_snaps.append((i, snap_time))
-        return_code = output.returncode
+        create_snapshot(db, cyhy_db_section, org_id, use_only_existing_snapshots=False)
 
-        if return_code == 0:
-            successful_snaps.append(i)
-            # TODO: If the org has children & their requestdoc has CYHY, add to descendant org list
-            successful_descendant_snaps += db.RequestDoc.get_all_descendants(i)
-            logging.info(
-                "Added to successful snaps: %s (%.2f s)", i, round(snap_time, 2)
+
+def generate_weekly_snapshots(db, cyhy_db_section):
+    """Generate all snapshots needed in order to generate the CyHy reports."""
+    start_time = time.time()
+
+    logging.info("Building list of reports to generate...")
+    reports_to_generate = create_list_of_reports_to_generate(db)
+
+    logging.info("Building list of snapshots to generate...")
+    snapshots_to_generate = create_list_of_snapshots_to_generate(
+        db, reports_to_generate
+    )
+
+    # List to keep track of our snapshot creation threads
+    snapshot_threads = list()
+
+    # Lists of orgs for each snapshot thread to process
+    snapshots_to_generate = list(
+        make_list_chunks(snapshots_to_generate, SNAPSHOT_THREADS)
+    )
+
+    # Start up the threads to create snapshots
+    for orgs in snapshots_to_generate:
+        try:
+            snapshot_thread = threading.Thread(
+                target=create_snapshots, args=(orgs, db, cyhy_db_section),
```

Comment:
I don't think it needs to be changed, but I don't think this is the most efficient way to parallelize this process.  Once a thread completes its assigned work it will sit idle.  Better would be to assign threads a single org at a time and give them new work to do (another org) when they are finished.  This way no threads would sit idle until there is no more work to be doled out.

Given that all threads are on the same instance in this case, there probably wouldn't be a lot of savings by rewriting this unless some threads are assigned groups of orgs whose snapshot creation takes considerably less time than others.  If we ever move to a Lambda or Kubernetes type of architecture for this processing, where there is heterogeneous hardware, then we would definitely want to rewrite the parallelization to handle a single org at a time.

Reply:
I added this info to the [cyhy-2.0 requirements issue](https://github.com/cisagov/cyhy-system/issues/7#issuecomment-759478386); this doesn't obviate creating an issue in this repo for this as a future improvement.

Reply:
I also created #62 for the parallelization improvement work.  I think it's good to mention it in both places.

Reply:
@jsf9k - I agree with you 100% here.  Thanks for creating #62.